### ISSUE TITLE
Report feature gate states via kubectl cluster-info dump

### ISSUE THREAD

**Issue by sftim** [kubernetes/kubernetes#87869]
**What would you like to be added**:
Improve Kubernetes so that I can run `kubectl cluster-info dump` and discover which feature gates are recognized by the cluster, as well as learn which components have which feature gates enabled.

**Why is this needed**:
As a cluster operator, I want to be able to discover the cluster's running configuration without having to probe for features or to look at the command line options that components are running with.
If I have access via `kubectl` only, I still want to be able to learn how the cluster is set up.

I think this is:
/sig cli
/sig api-machinery

**Comment by sftim** [comment]
Some background: https://kubernetes.slack.com/archives/C0EG7JC6T/p1576778805011700

**Comment by logicalhan** [comment]
+1

**Comment by neolit123** [comment]
are there plans to one day decouple kube components from the same list of feature gates?

**Comment by sftim** [comment]
> are there plans to one day decouple kube components from the same list of feature gates?
@neolit123 can you rephrase your question? I'm really sorry but I don't understand what you're asking.

**Comment by sftim** [comment]
Operators and controllers also have a need to discover what features are or aren't available, so that they can adjust behavior accordingly.

**Comment by neolit123** [comment]
> > are there plans to one day decouple kube components from the same list of feature gates?
> 
> @neolit123 can you rephrase your question? I'm really sorry but I don't understand what you're asking.

currently feature gates are a monolith defined here:
https://github.com/kubernetes/kubernetes/blob/652f52b51c05befece676f87d0ca8bafeded269d/pkg/features/kube_features.go#L27
and some of them do not make sense for all components.

my question is if in the future components can end up splicing the monolith list and taking FGs only relevant to them.

**Comment by lavalamp** [comment]
/priority awaiting-more-evidence

From my perspective the work required to do this is vastly larger than would be justified by the benefit.

> As a cluster operator, I want to be able to discover the cluster's running configuration without having to probe for features or to look at the command line options that components are running with.

Why?

> If I have access via kubectl only, I still want to be able to learn how the cluster is set up.

Why? Is this really true of everyone with access to kubectl?

Personally, I think the place to solve this problem is in the operator/environment running the control plane. E.g., if you run an app inside Kubernetes, everything is visible from the api: you can look at the pod to see the command line flags, env vars, secrets mounted, etc... The same would be true of the control plane if you ran it as a workload inside Kubernetes, either by self-hosting, stacking, or just permitting the master kubelets to register with the control plane and post the mirror pods.

**Comment by sftim** [comment]
@lavalamp let's say I have developers who don't have any privileged access to the hosts running the control plane.

Those people want to know how the cluster is set up. If I self-hosted the control plane then, yes, those colleagues can see the setup. However, I might not self-host it. I might not run the control plane as Pods. Also, my colleagues didn't want to care about those details. They just want to know what feature flags they can use.

**Comment by lavalamp** [comment]
It's not quite as bad if you literally just want the feature gates, usually people want to see everything :)

Really we should put them in a resource anyway; sending them as separate flags to each binary is just asking for mistakes. That's still a lot of work :(

**Comment by sftim** [comment]
If this information is readable via the Kubernetes API, it is more widely useful.

Some examples:
- a controller that wants to know whether to watch Endpoints or EndpointSlices
- a cloud-aware application that applies special configuration if the Sysctls feature gate is disabled
- a custom scheduler that adapts to TaintNodesByCondition
- a DaemonSet and its Pods that implements a logging helper and wants to know if CRIContainerLogRotation is in use
- an operator that adapts its behavior based on the HPAScaleToZero feature gate
- a controller that aborts unless SCTPSupport feature gate is enabled

**Comment by sftim** [comment]
I'd still like this, so:
/remove-lifecycle rotten

Hope that's OK.

**Comment by liggitt** [comment]
> Really we should put them in a resource anyway; sending them as separate flags to each binary is just asking for mistakes. That's still a lot of work :(

I'm on the fence here... consistency is generally good. Instant rollout of a settings change across all components is sometimes exactly what you want and is sometimes definitely not what you want. Imagine all kube-proxy instances on all nodes flipping behavior instantly on a configmap change... a bad config push would take down the network on all nodes.

**Comment by sftim** [comment]
For a cluster-wide implementation that doesn't imply all-at-once changes then, off the top of my head, how about either a ConfigMap of _cluster-scope_ defaults, or its own resource?

You can then put your own mechanism on top to, eg, omit a certain setting from the cluster wide defaults, manage that separately, and maybe one day when all the places are again consistent you go back to configuring a cluster scoped default (and your logs get a tiny bit less noisy).

I'm pretty neutral as to mechanism; the outcome I want is in the issue description and there are certainly lots of ways to get there. It _is_ going to be important to give a relevant answer when there isn't a single value for something across the cluster.

What would people expect to see in a cluster-info dump if, for example, `TTLAfterFinished=true` and also `TTLAfterFinished=false`, depending on which cluster controller manager you look at?

**Comment by keifgwinn** [comment]
I'm interested in this as well, especially when you're taking managed kubernetes from a provider.

/remove-lifecycle rotten

**Comment by pacoxu** [comment]
After checking code and website informations, the feature gates status is not that clear. Some mistakes in code comments or websites. I opened https://github.com/kubernetes/website/pull/26607 and 

However, we can write some scripts to calculate it from the client-side.  

- apiserver version; kubelet version; pods' image versions(pod/ds/deploy Yamls). (use regex to check feature gate args)
- read feature gates status from specific codes like https://github.com/kubernetes/kubernetes/blob/release-1.19/pkg/features/kube_features.go [support only latest 3-4 versions, make a yaml/CSV/map to manage them according to codes]

It may not be that accurate but would help in investigating issues.

**Comment by sftim** [comment]
With feature gates that are configurable (eg enabled by default, but you can disable it), this command should report the _actual_ status of feature gates. Making a guess is likely to mislead users.

**Comment by sftim** [comment]
Thinking about this again, maybe this could be done as a CRD or APIService that does something similar to the built-in (and deprecated) ComponentStatus API.
ComponentStatus was deprecated because it didn't fit as a core feature. However, the same functionality might work as an add-on, and kubectl could be aware of whether or not that API exists.

The trouble is: yes, that's making great use of Kubernetes' architecture and its built-in extensibility. It's also a _lot_ of work for a niche feature.

**Comment by pacoxu** [comment]
> Add `kubernetes_feature_enabled` metric series to track whether each active feature gate is enabled.

https://github.com/kubernetes/kubernetes/pull/112690 was merged in v1.26 cycle. (I just got to read the release note of this PR). It can fix partially on this issue and I wondered if we can add the metrics info into `kubectl cluster-info dump`

**Comment by pacoxu** [comment]
In cluster of v1.26+, we can use below command to get the feature enabled status.
```
kubectl get --raw /metrics  | grep kubernetes_feature_enabled
```
